### PR TITLE
First Deliverable - Eduardo Barrera

### PR DESCRIPTION
**Pull request by Aeross**
# Bootcamp first deliverable
Golang Bootcamp challenge

To run the project a config file is required. I'll share it separately.

Clone the project, navigate to the project's root directory and execute

``go run main.go``

the dependencies will be downloaded automatically.

3 endpoints will be available: 

GET ``/champions``
GET ``/champions/{id}``
POST ``/champion``

The first endpoint will return the first 10 elements in the database. Optionally, you can send a ``limit`` (/champions?limit=N) param to limit the db search to ``N`` elements where ``N`` is a positive integer.

The second endpoint will return a single record identified by ``{id}`` where ``{id}`` is a positive integer. The max amount of ids at the moment is **23**. Any number above the max amount of records in the database will simply return ``Not Found``

The third endpoint will insert a new record in the database.

i.e: 

GET to ``localhost:4000/champions``

returns:

```
{
    "name": "Draven",
    "lore": "In Noxus, warriors known as Reckoners face one another in arenas where blood is spilled and strength tested—but none has ever been as celebrated as Draven. A former soldier, he found that the crowds uniquely appreciated his flair for the dramatic, and his unparalleled skill with his spinning axes. Addicted to the spectacle of his own brash perfection, Draven has sworn to defeat whomever he must to ensure that his name is chanted throughout the empire forever more.",
    "created_at": "2020-11-22T04:31:50Z"
}
{
    "name": "Fiddlesticks",
    "lore": "Something has awoken in Runeterra. Something ancient. Something terrible. The ageless horror known as Fiddlesticks stalks the edges of mortal society, drawn to areas thick with paranoia where it feeds upon terrorized victims. Wielding a jagged scythe, the haggard, makeshift creature reaps fear itself, shattering the minds of those unlucky enough to survive in its wake. Beware the sounding of the crow, or the whispering of the shape that appears <i>almost</i> human… Fiddlesticks has returned.",
    "created_at": "2020-11-22T04:33:04Z"
}
.
.
.
.
```

GET to ``localhost:4000/champions/22``

returns:

```
{
    "name": "Thresh",
    "lore": "Sadistic and cunning, Thresh is an ambitious and restless spirit of the Shadow Isles. Once the custodian of countless arcane secrets, he was undone by a power greater than life or death, and now sustains himself by tormenting and breaking others with slow, excruciating inventiveness. His victims suffer far beyond their brief mortal coil as Thresh wreaks agony upon their souls, imprisoning them in his unholy lantern to torture for all eternity.",
    "created_at": "2020-11-22T16:21:59Z"
}
```
POST to ``localhost:4000/champion``

body:

```
{
    "name": "Ahri",
    "lore": "Innately connected to the latent power of Runeterra, Ahri is a vastaya who can reshape magic into orbs of raw energy. She revels in toying with her prey by manipulating their emotions before devouring their life essence. Despite her predatory nature, Ahri retains a sense of empathy as she receives flashes of memory from each soul she consumes."
}
```

returns:

``Success creating the Champion: N`` (where N is the id in the database)

Verify it with:

GET to ``localhost:4000/champions/N``

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	//Read config file
+	viper.SetConfigName("config")
+	viper.SetConfigType("json")
+	viper.AddConfigPath(".")
+	err := viper.ReadInConfig()
+
+	if err := viper.ReadInConfig(); err != nil {
+		if _, ok := err.(viper.ConfigFileNotFoundError); ok {
+			// Config file not found
+			panic(fmt.Errorf("fatal error config file not found: %s", err))
```
Comment (RolloCasanova) [inline config/config.go:32]: Better to handle all panics at the top level instead of here.

Your function returns an error; use that at your favor.

File: config/config.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	dbPort := viper.GetString("db.port")
+	dbName := viper.GetString("db.name")
+	dsn := fmt.Sprintf("%s:%s@(%s:%s)/%s?parseTime=true", username, password, dbHost, dbPort, dbName)
+
+	// Open db connection
+	db, err := openDB(dsn)
+	if err != nil {
+		errorLog.Fatalf("message: unable to open db connection, type: database, err: %v", err)
+		return nil, err
+	}
```
Comment (RolloCasanova) [inline config/config.go:59]: IMHO, you should only open the connection when using it, and close it after using it.

Also, I think you are mixing responsibilities here. The idea of the config is to have the configuration ready for when you are going to use it, not to initialize this kind of connection.

This (alongside moving the `OpenDB` function) should be moved/done at the repositories level, just before making the corresponding use of DB.

File: config/config.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+	// Retrun the config
+	return &Config{dsn, infoLog, errorLog, db, &addr}, nil
```
Comment (RolloCasanova) [inline config/config.go:62]: If you agree with the previous comment, please remove `db` from the Config struct, as that's not the right place for it.

File: config/config.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	dsn := fmt.Sprintf("%s:%s@(%s:%s)/%s?parseTime=true", username, password, dbHost, dbPort, dbName)
+
+	// Open db connection
+	db, err := openDB(dsn)
+	if err != nil {
+		errorLog.Fatalf("message: unable to open db connection, type: database, err: %v", err)
+		return nil, err
+	}
+
+	// Retrun the config
```
Comment (RolloCasanova) [inline config/config.go:61]: typo 🌚

File: config/routes.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
+package config
+
+import (
+	"net/http"
+
+	"github.com/wizelineacademy/golang-bootcamp-2020/domain/repositories"
+
+	"github.com/wizelineacademy/golang-bootcamp-2020/interface/handlers"
+
+	"github.com/gorilla/mux"
```
Comment (RolloCasanova) [inline config/routes.go:10]: By convention, imported packages in a file should be separated into three separated blocks:

1st block: Golang library packages
2nd block: Your project's own dependencies (local packages)
3rd block: 3rd party packages

```suggestion
	"net/http"

	"github.com/wizelineacademy/golang-bootcamp-2020/domain/repositories"
	"github.com/wizelineacademy/golang-bootcamp-2020/interface/handlers"

	"github.com/gorilla/mux"
```

File: config/routes.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package config
+
+import (
+	"net/http"
+
+	"github.com/wizelineacademy/golang-bootcamp-2020/domain/repositories"
```
Comment (RolloCasanova) [inline config/routes.go:6]: By convention, imported packages in a file should be separated into three separated blocks:

1st block: Golang library packages
2nd block: Your project's own dependencies (local packages)
3rd block: 3rd party packages

```suggestion
	"net/http"

	"github.com/wizelineacademy/golang-bootcamp-2020/domain/repositories"
	"github.com/wizelineacademy/golang-bootcamp-2020/interface/handlers"

	"github.com/gorilla/mux"
```

File: domain/repositories/champion.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+	// Copy the values from each field in sql.Row to the corresponding field in the Champion struct.
+	err := row.Scan(&c.Name, &c.Lore, &c.DateCreated)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, ErrNoRecord
+		} else {
+			return nil, err
+		}
+	}
```
Comment (RolloCasanova) [inline domain/repositories/champion.go:51]: You could get rid of the whole else block and only leave the return, as your code (at this moment) doesn't have any other possible output.

```suggestion
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, ErrNoRecord
		} 
		return nil, err
	}
```

File: domain/repositories/champion.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+// NewChampRepo returns an inituialized ChampRepo struct
+func NewChampRepo(DB *sql.DB) *ChampRepo {
+	return &ChampRepo{DB}
+}
+
+// GetSingle gets a single database row and returns it as a Champion
+func (cr *ChampRepo) GetSingle(id int) (*models.Champion, error) {
+	// SQL statement
+	stmt := `SELECT  name, lore, created FROM champions WHERE id = ?`
```
Comment (RolloCasanova) [inline domain/repositories/champion.go:35]: This should be on the Repository, and as it won't change, you should declare it as a constant.

This applies to the other statement on the other function.

As you may see, you are creating it here, and just passing it to the next layer. Why do that if you could simply use it right on the repository layer?

File: domain/repositories/champion.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+		c := &models.Champion{}
+
+		// Use row.Scan() to copy the values from each field in sql.Row to the corresponding field in the Champion struct.
+		err := rows.Scan(&c.Name, &c.Lore, &c.DateCreated)
+		if err != nil {
+			return nil, err
+		}
+
+		champions = append(champions, c)
+
```
Comment (RolloCasanova) [inline domain/repositories/champion.go:83]: Please, remove this empty line.

File: domain/repositories/champion.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+	}
+
+	// Get the ID of the newly inserted record.
+	id, err := result.LastInsertId()
+	if err != nil {
+		return 0, err
+	}
+
+	// The ID returned has the type int64, so we convert it to an int type before returning it.
+	return int(id), nil
```
Comment (RolloCasanova) [inline domain/repositories/champion.go:112]: Why not use int64 as a return?

File: domain/repositories/champion.go [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+		}
+
+		champions = append(champions, c)
+
+	}
+
+	// When the rows.Next() loop has finished we call rows.Err() to retrieve any error that was encountered during the iteration.
+	if err = rows.Err(); err != nil {
+		return nil, err
+	}
```
Comment (RolloCasanova) [inline domain/repositories/champion.go:89]: No need to make an assignment here. This could be simplified as follows:
```suggestion
	if rows.Err() != nil {
		return nil, rows.Err()
	}
```

File: domain/repositories/champion.go [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+// ChampRepo defines the link between the Champion and the Database
+type ChampRepo struct {
+	DB *sql.DB
+}
+
+// NewChampRepo returns an inituialized ChampRepo struct
+func NewChampRepo(DB *sql.DB) *ChampRepo {
+	return &ChampRepo{DB}
+}
```
Comment (RolloCasanova) [inline domain/repositories/champion.go:30]: By convection, `DB` should be changed to `db`.
```suggestion
func NewChampRepo(db *sql.DB) *ChampRepo {
	return &ChampRepo{db}
}
```

File: domain/repositories/champion.go [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+	err := row.Scan(&c.Name, &c.Lore, &c.DateCreated)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, ErrNoRecord
+		} else {
+			return nil, err
+		}
+	}
+
+	//If everything went OK then return the Champion struct.
```
Comment (RolloCasanova) [inline domain/repositories/champion.go:53]: By Golang convention, comments should have a blank space between `//` and the first comment's character.
```suggestion
	// If everything went OK then return the Champion struct.
```

File: interface/handlers/champions.go [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+	err = c.ToJSON(w)
+	if err != nil {
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		ch.errorLog.Printf("message: error while parsing the champions data, handler: champions, method: GET, id: %v, err: %v\n", id, err)
+	}
+}
+
+// GetChamps returns multiple Champions. It defaults to 10 if a limit isn't specified or if the limit is < 1.
+func (ch *ChampionHandler) GetChamps(w http.ResponseWriter, r *http.Request) {
+
```
Comment (RolloCasanova) [inline interface/handlers/champions.go:59]: Please, remove this empty line.

File: interface/handlers/champions.go [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+	const (
+		defaultLimit = 10
+	)
```
Comment (RolloCasanova) [inline interface/handlers/champions.go:62]: Single const definition shouldn't be a const block.
```suggestion
	const defaultLimit = 10
```

File: interface/handlers/champions.go [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+
+	// Marshal the Champions slice to JSON
+	for _, c := range champions {
+		err = c.ToJSON(w)
+		if err != nil {
+			ch.errorLog.Printf("message: error while parsing the champions data, handler: champions, method: GET, err: %v\n", err)
+			helpers.ServerError(w, ch.errorLog, err)
+		}
+	}
+
```
Comment (RolloCasanova) [inline interface/handlers/champions.go:103]: Please, remove this empty line.

File: interface/handlers/champions.go [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+	// ID of the new record back.
+	id, err := ch.champRepo.Insert(champ)
+	if err != nil {
+		fmt.Printf("err %v\n", err)
+		helpers.ServerError(w, ch.errorLog, err)
+		return
+	}
+
+	fmt.Fprintf(w, "Success creating the Champion: %v\n", id)
+
```
Comment (RolloCasanova) [inline interface/handlers/champions.go:127]: Please, remove this empty line.

File: main.go [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+	defer config.DB.Close()
+
+	sm := config.InitRoutes()
+	// HTTP Server configuration
+	s := &http.Server{
+		Addr:         *config.Addr,      //set the bind address
+		Handler:      sm,                //set the default handler
+		ReadTimeout:  5 * time.Second,   //max time to read request from the client
+		WriteTimeout: 10 * time.Second,  //max time to write request from the client
+		IdleTimeout:  120 * time.Second, //max time for connections using TCP Keep-Alive
```
Comment (RolloCasanova) [inline main.go:33]: 5, 10, and 120 are magic numbers.

Define them as part of the configuration (or as constants).

File: main.go [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+	sigChan := make(chan os.Signal)
+	signal.Notify(sigChan, os.Interrupt)
+	signal.Notify(sigChan, os.Kill)
+
+	// If the server was interrupted, print the interruption signal that stopped it.
+	sig := <-sigChan
+	infoLog.Println("Received terminate, graceful shutdown", sig)
+
+	// Wait until the server has finished processing pending jobs, then terminate. Or, if the time threshold is met, terminate.
+	tc, _ := context.WithTimeout(context.Background(), 30*time.Second)
```
Comment (RolloCasanova) [inline main.go:59]: 30, another magic number.

Also, In this specific case, context.WithTimeout returns a cancel function.

Reading about it, this cancel function should be called in order to avoid a context leak.

File: interface/handlers/champions.go [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+		if paramLimit < 1 {
+			ch.infoLog.Printf("message: limit < 1, using default limit %v, handler: champions, method: GET", defaultLimit)
+			limit = defaultLimit
+		} else {
+			limit = paramLimit
+		}
+	} else { // If query param "limit" isn't found
+		limit = defaultLimit
+	}
```
Comment (RolloCasanova) [inline interface/handlers/champions.go:83]: A little rearangement of this block, what do you think?:
```suggestion
	limit := defaultLimit
	params := r.URL.Query()
	// If query param "limit" is found
	if params.Get("limit") != "" {
		paramLimit, err := strconv.Atoi(params.Get("limit"))
		if err != nil {
			ch.errorLog.Printf("message: error parsing limit, handler: champions, method: GET, err: %v\n", err)
			helpers.NotFound(w)
			return
		}

		if paramLimit < 1 {
			ch.infoLog.Printf("message: limit < 1, using default limit %v, handler: champions, method: GET", defaultLimit)
		} else {
			limit = paramLimit
		}
	}
```

**Comment by RolloCasanova** [comment]
Please decline this PR, and add the corresponding changes to your own project.

File: config/config.go (hunk 2 above)
Comment (Aeross) [reply config/config.go:59]: According to:

https://golang.org/pkg/database/sql/#OpenDB

> the OpenDB function should be called just once. It is rarely necessary to close a DB.

And according to a book called "Let's Go!" by Alex Edwards:

> The sql.Open() function returns a sql.DB object. This isn’t a database connection — it’s a pool of many connections. Go manages these connections as needed, automatically opening and closing connections to the database via the driver.
> 
> The connection pool is intended to be long-lived. In a web application it’s normal to initialize the connection pool in your main() function and then pass the pool to your handlers. You shouldn’t call sql.Open() in a short-lived handler itself — it would be a waste of memory and network resources.

Therefore, as the connection is required to be initialized just once, at the time I was writing the code, it made sense to me to put it in a the configuration file and inject it as a dependency. However, I agree that I can move it to the repositories level as well and initialize it there as it's part of it.

What do you think?

File: main.go (hunk 19 above)
Comment (Aeross) [reply main.go:59]: Ah, I forgot to remove the underscore and handle that cancel. Noted.
